grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerQueryImpl.java

```diff
@@ -598,4 +601,17 @@ public QueryState subGoal(Answer sub, Unifier u, QueryState parent, Set<Reasoner
         return Sets.cartesianProduct(atomOptions).stream()
                 .map(atomList -> ReasonerQueries.create(new HashSet<>(atomList), graph()));
     }
+
+    /**
+     * reiteration might be required if rule graph contains loops with negative flux
+     * or there exists a rule which head satisfies body
+     * @return true if because of the rule graph form, the resolution of this query may require reiteration
+     */
+    public boolean requiresReiteration() {
+        RuleGraph ruleGraph = new RuleGraph(graph());
+
+        Set<InferenceRule> dependentRules = ruleGraph.getDependentRules(this).collect(Collectors.toSet());
+        return ruleGraph.subGraphHasLoopsWithNegativeFlux(dependentRules)
```

Comment:
[*] - This could potentially be 
```
RuleUtil. subGraphHasLoopsWithNegativeFlux(graph())` || `RuleUtil. subGraphHasRulesWithHeadSatisfyingBody(graph())
```

Again, no object allocation needed